fix(docs-infra): improve accessibility of aio-select component

dario-piotrowicz opened the pull request:
improve the accessibility of the aio-select component so that it is
clear for screen reader users its functionality (currently it is
presented as a simple button), following the WAI-ARIA authoring
practices (see: https://www.w3.org/TR/wai-aria-practices/#combobox)

A first attempt in improving the accessibility of the component has been
tried in PR #45937 by replacing it with the material select component,
such implementation has however been scrapped since the increase of
payload sizes has proven prohibitively large

(also note that given native select elements haven't been used given the lack
of syling options for such elements)

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A

Screen readers read the aio-select elements as if they were normal buttons without clearly indicating to users how to interact with them:

https://user-images.githubusercontent.com/61631103/168682930-57d923bf-ac60-40f5-b315-c3ddfbf42072.mp4

(note in the video how the elements are referred as normal buttons , note that the keyboard interaction is pretty ad-hoc and does not follow the wai-aria authoring practices)

## What is the new behavior?

both aspects have been improved making the element more accessible:

https://user-images.githubusercontent.com/61631103/168683162-afa33cf4-b4b5-4cbf-bf5b-a21570f9ff5c.mp4

(note in the video how the elements are referred as combo boxes now, note that the keyboard interaction now follows the wai-aria authoring practices)


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information
 - @gkalpak :slightly_smiling_face: 
 - I wanted to re-implement this using the native select elements. but I was unaware/forgot that such type of element can't really be styled :see_no_evil: (I guess I now know why this component was created in the first place) so I just kept it as is but improved its a11y
 - I did add unit tests for keyboard navigation, I thought of adding unit tests for aria attributes as well, but that felt a bit overkill, but I do can add them if we deem it beneficial :slightly_smiling_face:

dario-piotrowicz committed:
fix(docs-infra): improve accessibility of aio-select component

improve the accessibility of the aio-select component so that it is
clear for screen reader users its functionality (currently it is
presented as a simple button), following the WAI-ARIA authoring
practices (see: https://www.w3.org/TR/wai-aria-practices/#combobox)

A first attempt in improving the accessibility of the component has been
tried in PR #45937 by replacing it with the material select component,
such implementation has however been scrapped since the increase of
payload sizes has proven prohibitively large

(also note that given native select elements haven't been used given the lack
of syling options for such elements)

dario-piotrowicz commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -17,13 +17,34 @@ export class SelectComponent implements OnInit {
+    }
+  }
+
+  listBoxId = `aio-select-list-box-${Math.floor(Math.random() * 10000)}`;
```
inelegant way to get unique ids here :sweat_smile: , please let me know if there's a better alternative

gkalpak replied:
I think it is reasonable to have a simple counter (either as a static property or as a variable outside the class) and increment it for every instance we create.

dario-piotrowicz replied:
ah yeah good idea! :smiley::+1:

dario-piotrowicz replied:
I've added a static property to the class :slightly_smiling_face::+1:

dario-piotrowicz commented on aio/src/app/app.component.ts:
```diff
@@ -477,4 +477,13 @@ export class AppComponent implements OnInit {
+    // call (e.g. the aio-select expanding) the scroll will happen after the DOM has
+    // been updated
+    setTimeout(() =>
+      this.sidenav._content.nativeElement.scrollTo({top: window.innerHeight, behavior: 'smooth'})
```
small improvement I tried to make for the versions select on the sidenav, basically currently you open it and need to scroll down in order to see the elements:
![old](https://user-images.githubusercontent.com/61631103/168684836-a90d403c-fdb4-48c5-963a-a13d8c3bc497.gif)
With the new implementation it scrolls by itself:
![new](https://user-images.githubusercontent.com/61631103/168684871-434dc4f7-4a47-4176-86f6-d24ff9bfeea3.gif)

(I just noticed that it doesn't scroll all the way down... I wonder if I need to increase the add a small time to the `setTimeout` or something :thinking: )

Is this a good idea? :slightly_smiling_face:

dario-piotrowicz replied:
ah yeah! you're totally right! :smiley: 

I've applied the change, I am getting the select element using `ViewChild` let me know if you think it is ok (since the api list comes from the docs I'd imagine that the aio-select in the sidenav will always be the first one, do you think that it is ok to make such assumption?)

Another thing I wanted to mention is that I have seen that `scrollIntoView` has some experimental options I was tempted to use them for the smooth behaviour, what do you think? is it better to keep it simple as it currently is?

Ah and an extra note, I intend to try to change the doc version select into something else (I haven't actually yet thought that through anyways :sweat_smile:) at some point after this PR (in order to resolve https://github.com/angular/angular/issues/44339) so the code here may end up being removed anyways :slightly_smiling_face:

gkalpak replied:
Using smooth scrolling would be file as long as it is silently ignored in non-supporting browsers (which I believe it is).
Up to you :smiley:

dario-piotrowicz replied:
good point :slightly_smiling_face::+1:

mary-poppins commented:
You can preview 34ad221 at https://pr46013-34ad221.ngbuilds.io/.

gkalpak commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -17,13 +17,34 @@ export class SelectComponent implements OnInit {
   // eslint-disable-next-line @angular-eslint/no-output-native
   @Output() change = new EventEmitter<{option: Option, index: number}>();
 
+  @Output() optionToggled = new EventEmitter<boolean>();
```
`optionToggled` sounds confusing :smiley:
Did you mean `optionsToggled`?

dario-piotrowicz replied:
yeah sorry :sweat:

gkalpak commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -53,8 +75,35 @@ export class SelectComponent implements OnInit {
+      }
+    }
+    switch(event.key) {
+      case 'ArrowDown':
+        runOrOpenOptions(() => this.currentOptionIdx = Math.min(this.currentOptionIdx + 1, (this.options?.length ?? 0) - 1));
+        break;
+      case 'ArrowUp':
+        runOrOpenOptions(() => this.currentOptionIdx = Math.max(this.currentOptionIdx - 1, 0));
+        break;
```
OOC, why not loop back to the beginning/end?

Also, is it a common ARIA pattern that pressing the up/down arrow while the options are hidden will show the options?

dario-piotrowicz replied:
Yes, again it comes from the same [wai aria example](https://www.w3.org/TR/wai-aria-practices/examples/combobox/combobox-select-only.html) I mentioned in the other comment :slightly_smiling_face: 

![Screenshot at 2022-05-18 21-35-02](https://user-images.githubusercontent.com/61631103/169151072-d956c91b-0d20-4770-830f-68c004044370.png)

it is also what the native select does

(PS: note that in the text there the visual focus is different from the DOM focus, that one does not move :slightly_smiling_face:)

gkalpak replied:
Cool, thx for the info :+1:

gkalpak commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -53,8 +75,35 @@ export class SelectComponent implements OnInit {
+        runOrOpenOptions(() => this.select(this.currentOptionIdx));
+        break;
+    }
+    if(event.code !== 'Tab') {
```
Is it a common ARIA pattern that pressing `Tab` while the options are shown will get you to the next element (outside the combobox) instead of jumping to the next option?

dario-piotrowicz replied:
Yes, as you can see from this example from the wai aria practices: https://www.w3.org/TR/wai-aria-practices/examples/combobox/combobox-select-only.html

![Screenshot at 2022-05-18 19-43-57](https://user-images.githubusercontent.com/61631103/169149280-c7a5183e-429e-4742-8424-2b6d34daad8e.png)

(Actually I just noticed that tab should also select the element, I need to add that :see_no_evil:)

That's also (almost) the same behavior as the native select element :slightly_smiling_face: (if you prefer the native behavior I can apply that instead)

I am pretty sure that it is because the select is considered a single focusable control and the options are simple options for it, so with tab you want to move on the next focusable control rather then through the options of the current one (it is for example the same as with radio buttons, in a group tab doesn't move focus on the next button but to the next control after the group of buttons, I think it's the same sort of logic :slightly_smiling_face: )

(this is indeed something that before in the component worked quite ad-hoc in my opinion since tabbing was actually the only way to move between options)

gkalpak replied:
Makes sense. Thx!

gkalpak reviewed (changes requested):
Thx for working on this, @dario-piotrowicz, (and for investigating the various options).

dario-piotrowicz commented on aio/src/app/shared/select/select.component.spec.ts:
```diff
@@ -140,12 +125,88 @@ describe('SelectComponent', () => {
+      fixture.detectChanges();
+    };
+
+    const printK = (key:string) => key === ' ' ? "' '" : key;
```
I was trying to save 2 precious keystrokes :see_no_evil:

dario-piotrowicz replied:
(and actually above we've got `pressKey` :facepalm: sorry :sweat:)

dario-piotrowicz commented on aio/src/app/shared/select/select.component.html:
```diff
@@ -1,16 +1,33 @@
+    (keydown)="handleKeydown($event)"
+    [attr.aria-expanded]="showOptions"
+    [attr.aria-activedescendant]="currentOptionIdx > -1 && showOptions ? listBoxId + '-option-' + currentOptionIdx : null"
+    [attr.aria-label]="(label || '') +  (selected?.title || '')"
```
sorry I thought that the nullish operator was not supported in templates, seems like it's been supported for a while :see_no_evil:

dario-piotrowicz commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -53,8 +75,35 @@ export class SelectComponent implements OnInit {
+      if(!this.showOptions) {
+        this.showOptions = true;
+        const indexOfSelected =  (this.options ?? []).indexOf(this.selected);
+        this.currentOptionIdx = indexOfSelected > 0 ? indexOfSelected : 0;
```
yes it is persistent, but indeed you may not want to use the persisted value, see here for example:

![Peek 2022-05-18 21-50](https://user-images.githubusercontent.com/61631103/169153777-8fe65818-9011-4dd3-9991-09359765aa46.gif)

I select Const then open the options and navigate to Enum, but I don't select it (I press `escape` which only closes the list), I move away and when I come back I would expect to start back from Const, but that doesn't happen if we use the persisted value, we'll just be moved to Enum, which feels a bit sluggish to me :sweat_smile:

gkalpak replied:
Oh, I see. I mixed "current" with "selected". 
Thx for clarifying :+1:

dario-piotrowicz commented on aio/src/app/shared/select/select.component.ts:
```diff
@@ -17,13 +17,34 @@ export class SelectComponent implements OnInit {
   @Input() showSymbol = false;
 
-  @Input() label: string;
+  @Input() label?: string;
```
@gkalpak I needed to add this otherwise we'd get this warning in the template:
![Screenshot at 2022-05-18 22-06-51](https://user-images.githubusercontent.com/61631103/169156989-81ec37b8-e5d4-42d8-a46c-80f56ac474ae.png)

gkalpak replied:
Actually, since we initialize it in [ngOnInit()][1], I think it is OK to use `!` instead (and not have to `?? ''` in the template.

[1]: https://github.com/angular/angular/blob/5ced1deb8fa4c659e12a885ca1b0db7191dd0c2d/aio/src/app/shared/select/select.component.ts#L54

dario-piotrowicz replied:
ah dang! yeah! sorry I missed that! :sweat::+1:

dario-piotrowicz commented on aio/src/app/shared/select/select.component.spec.ts:
```diff
@@ -140,12 +125,88 @@ describe('SelectComponent', () => {
+        pressKey(key);
+        expect(host.onChange).toHaveBeenCalledWith({ option: options[0], index: 0 });
+        expect(getButton().textContent).toContain(options[0].title);
+        expect(getButtonSymbol()?.className).toContain(options[0].value);
```
I've added `not.toContain` checks before the `pressKey` line, is that ok? :slightly_smiling_face:

gkalpak replied:
:+1:

dario-piotrowicz commented:
@gkalpak thanks a lot for the awesome (as always) reviewing!!! :heart_eyes:

I've replied to your comments please have another look :slightly_smiling_face:

(if you could also re-kick the ci/cd that'd be great :pray:)

gkalpak commented on aio/src/app/app.component.ts:
```diff
@@ -477,4 +481,12 @@ export class AppComponent implements OnInit {
+
+  scrollSelectIntoView() {
+    // this method is used to scroll the select component into view when it is clicked/opened
+    // the setTimeout is needed so that the scroll happens after the component has expanded
```
Why do we need the scroll to happen after it has expanded?

dario-piotrowicz replied:
without it the scrolling doesn't actually work, the fact is that the non-expanded aio-select element doesn't actually need any scrolling (it is already there at the bottom but fully visible, also there is nothing below it, so it cannot scroll up top anyways) the only way to actually have the scrolling is to first have the component expanded (so that there is some scrolling available to be done)

dario-piotrowicz replied:
![Peek 2022-05-20 20-30](https://user-images.githubusercontent.com/61631103/169598803-a1e3c3f3-c092-4fdc-92fb-10c6fc9fb4dc.gif)

mary-poppins commented:
You can preview b0e25d4 at https://pr46013-b0e25d4.ngbuilds.io/.

gkalpak reviewed (commented):
Now that CI ran, can you please make tests pass, @dario-piotrowicz 🙏

dario-piotrowicz commented:
> Now that CI ran, can you please make tests pass, @dario-piotrowicz pray

Sure, thanks a lot!!! :slightly_smiling_face:

dario-piotrowicz commented:
@gkalpak tests fixed, I just needed to update the aio payloads :slightly_smiling_face: 

there was also a strange failure:
![Screenshot at 2022-05-23 19-06-00](https://user-images.githubusercontent.com/61631103/169880552-d6eb9276-79d0-47ee-a43b-148bc64ec603.png)

But it is unrelated to any of the files I changed and also the suggested command didn't modify any file on my machine

:shrug:

gkalpak commented:
The strange failure should be fixed by #46082.
Can you rebase on latest `main` (and update `aio-payloads.json` again if needed)? :pray:

dario-piotrowicz commented:
> The strange failure should be fixed by #46082. Can you rebase on latest `main` (and update `aio-payloads.json` again if needed)? pray

great thanks! :slightly_smiling_face::+1:

by the way, I've rebased (and squashed all the commits into one, sorry for the inconvenience :sweat:)

sorry I am not sure how to check the sizes locally, usually I rely on the ci for that :cry: (which seems still not to work properly :sweat:)

mary-poppins commented:
You can preview 3f3944a at https://pr46013-3f3944a.ngbuilds.io/.

dario-piotrowicz committed:
fixup! fix(docs-infra): improve accessibility of aio-select component

update aio payload sizes

dario-piotrowicz commented:
Payloads update! thanks @gkalpak :heart_eyes:

gkalpak commented:
> closing and reopening the PR is not all it takes to make the ci run, is it? 🤔

It pretty much is (most of the time) :grin:
We/CircleCI are looking into the issue, but for the time being...

gkalpak reviewed (approved):
Thx, @dario-piotrowicz :+1:
The size increase is not great, but worth the a11y improvements 💪

dario-piotrowicz commented:
> > closing and reopening the PR is not all it takes to make the ci run, is it? thinking
> 
> It pretty much is (most of the time) grin We/CircleCI are looking into the issue, but for the time being...

awesome! I'll keep that in mind! thanks! :grin:

dario-piotrowicz commented:
> Thx, @dario-piotrowicz +1 The size increase is not great, but worth the a11y improvements muscle

Thanks! :blush: (and sorry for the increase :cry:)

jessicajaniuk reviewed (approved):
reviewed-for: size-tracking

jessicajaniuk commented:
This PR was merged into the repository by commit 974da3f8026cfdd41fb1af2019355521add79122.
